terraformize/terraformize_terraform_wrapper.py

```diff
@@ -1,32 +1,38 @@
 from python_terraform import *
 from typing import Tuple, Optional
+import tempfile
+import shutil
 
 
 class Terraformize:
 
-    def __init__(self, workspace: str, folder_path: str, terraform_bin_path: Optional[str] = None):
+    def __init__(self, remote_backend: bool, workspace: str, folder_path: str, terraform_bin_path: Optional[str] = None):
         """
         Will create a terraform object, create a workspace & init the terraform directory
 
         Arguments:
+            :param remote_backend: if terraform is running with a remote state backend
             :param workspace: the workspace terraform will be executed in
             :param folder_path: the full path of the folder to run the terraform in
-            :param terraform_bin_path: the full path of the terraform binary to use, will try to use the one at the path
-            if not set
+            :param terraform_bin_path: the full path of the terraform binary to use, defaults to using $PATH if unset
         """
         self.terraform_bin_path = terraform_bin_path
-        self.tf = Terraform(working_dir=folder_path, terraform_bin_path=self.terraform_bin_path)
 
-        # we need to init for a remote backend before we can create a workspace
-        self.init_return_code, self.init_stdout, self.init_stderr = self.tf.init(dir_or_plan=folder_path)
+        if remote_backend is True:
+            self.working_path = tempfile.mkdtemp()
+            self.tf = Terraform(working_dir=self.working_path, terraform_bin_path=self.terraform_bin_path)
+            # using an envvar is the only way to prevent terraform from creating
+            # a default.tfstate during init on a remote backend (which causes locking)
+            os.environ["TF_WORKSPACE"] = workspace
+            self.init_return_code, self.init_stdout, self.init_stderr = self.tf.init(from_module=folder_path, dir_or_plan=self.working_path)
+        else:
+            self.tf = Terraform(working_dir=folder_path, terraform_bin_path=self.terraform_bin_path)
+            self.init_return_code, self.init_stdout, self.init_stderr = self.tf.init(dir_or_plan=folder_path)
+            # create workspace and if failed we assume it already exists
+            self.tf.create_workspace(workspace=workspace)
+            self.workspace_return_code, self.workspace_stdout, self.workspace_stderr = self.tf.set_workspace(workspace=workspace)
 
-        # always create the workspace and switch to it, if workspace already created carries on when it failed creating
-        # it again
-        self.tf.create_workspace(workspace=workspace)
```

Comment:
As per our talk at #18 we're sticking to file based workspace as the TF_WORKSPACE envvar doesn't guarantee creation of workspace on all backends (only select of pre-existing workspaces)

Reply:
file based doesnt work properly with remote backends - i kept it that way for local backends though. 

with remote backends you have to perform an `init` before a `workspace create`. but with no possibility to select a workspace yet, it will create a `default.tfstate` in the remote backend. this is a problem with concurrency because N parallel api calls will trigger them all to try and create a `default.tflock` and cause all but 1 api call to fail. using the `TF_WORKSPACE` envvar is the only way to prevent creating a `default.tfstate` and allows concurrency.

Reply:
The problem with that is there's no guarantee that all backends allow creation of workspaces via envvar but I do have a somewhat creative solution to the problem... what if we run file based & at the init stage have it wrapped with a try/except block so that if it fails due to another API call creating the default.tflock we know we can skip that step & just carry on towards the workspace creation & switch (it's my understanding that only the original default init is shared... so once we get pass the original init the rest won't be problematic)

Reply:
if the init cant get a lock then it will fail to setup the local directory structure and we cant proceed with create/select of a workspace. so far s3/gcs/consul works with the envvar method.

Reply:
I'll create a test run with all currently known backends & see which are affected by the non creation of workspaces with the envvar, if we get lucky & it turns out non are (despite it being mentioned otherwise on the terraform repo) I'll merge this in

Reply:
The more I read up about it the more it's clear to me that TF_WORKSPACE does not create workspace by design (while they are some backends that create it this isn't something we can take for granted across the board) so the workspace creation part will have to be handled via file based, I'm fine with everything after the workspace creation using the TF_WORKSPACE envvar but for the creation part we'll need to figure out a different solution (even if it is just a retry loop until it gets the init stage unblocked)

Reply:
understood. hopefully you can come up with a solution that scales. i need this to be high throughput concurrency and the only way for that (atm) is with TF_WORKSPACE. i realize this may not be the average consumer use case though.

you could set `terraform init -lock-timeout=0s` to a value of 30s or something. but this is like banging your head against the door until it opens. it doesnt truely solve concurrency issues. multiple API calls to the same module (regardless of workspace) will constantly clash with locking issues.

Reply:
you could also set `terraform init -lock=true` to false, but i dont know what unintended side effects that may have. you could end up with a correct `default.tfstate` at some point and then what happens? probably not ideal

Reply:
There's another option that might be a bit more work - have Terraformize support both file based & envvar based approches & give the enduser the choice. (with the default being the file based one).

that would negate any objection I have as it's just an option so by turning it on the enduser is aware of it working with the envvar creating the workspace